File: more-topics/easy-1/find_missing_numbers.rb
# Find Missing Numbers

# P
#   Given a sorted array of integers, return an array of all the integers missing
#   between the highest and the lowest.
#   Integers may be positive or negative. Each integer will only appear once.
# Examples below.
# Data: Arrays and Range
# A
#   Use the first and last elements of the array to define a range.
#   Create a new array, 'results'.
#   Iterate over the range, and add each number to results if it is not included
#     in the initial array.
#   Return results.
# C

def missing(array)
  results = []
  (array.first...array.last).each do |num|
    results << num unless array.include?(num)
  end
  results
end

p missing([-3, -2, 1, 5]) == [-1, 0, 2, 3, 4]
p missing([1, 2, 3, 4]) == []
p missing([1, 5]) == [2, 3, 4]
p missing([6]) == []

File: JS210/pp_arrays/find_missing_numbers.js
// Find Missing Numbers

// Given a sorted array of integers, return an array including all the
// missing integers between the first and last integers of the given array

// Functional style is gonna shine again
// now where'd that range go...
function range(len, from) {
  return [...Array(len).keys()].map(n => n + from);
}

// Okay, so we'll make a range and filter it based on a negated inclusion of the range.

const missing = array => range(array[array.length - 1] - array[0] + 1, array[0]).filter(n => !array.includes(n));

console.log('Test Results');
console.log(missing([-3, -2, 1, 5]));                  // [-1, 0, 2, 3, 4]
console.log(missing([1, 2, 3, 4]));                    // []
console.log(missing([1, 5]));                          // [2, 3, 4]
console.log(missing([6]));                             // []
